skrub/_join_aggregator.py

```diff
@@ -0,0 +1,354 @@
+        flatten_ops = []
+        for col in [*num_mode_cols, *categ_mode_cols]:
+            flatten_ops.append(
+                pl.col(col).list[0].alias(col)
+            )
+        return agg_table.with_columns(flatten_ops)
+          
+
+    def join(self, left, right, left_on, right_on):
+        
+        return left.join(
+            right,
+            how="left",
+            left_on=left_on,
+            right_on=right_on,
+        )
+    
+
+def check_cols(tables):
+    """Check that all columns to join and columns to aggregate 
+    belong to their respective dataframes.
+    """
+    for idx, (table, cols_to_join, cols_to_agg) in enumerate(tables, start=1):
+
+        cols_to_join = np.atleast_1d(cols_to_join).tolist()
+        cols_to_agg = np.atleast_1d(cols_to_agg).tolist()
+
+        table_cols = set(table.columns)
+        input_cols = set([*cols_to_join, *cols_to_agg])
+
+        missing_cols = input_cols - table_cols
+        if len(missing_cols) > 0:
+            raise ValueError(f"{missing_cols} are missing in table {idx}")
+
+    return
+    
+
+class JoinAggregator(BaseEstimator, TransformerMixin):
+    """Perform aggregation on auxilliary dataframes before joining
+    on the base dataframe.
+
+    Apply numerical (mean, std, min, max) and categorical (mode) aggregation 
+    operations on the columns to agg, selected by dtypes.
+    
+    The grouping columns used during the aggregation are the columns used 
+    as keys for joining.
+
+    These operations can run lazily by inputing polars LazyFrames. 
+    Pandas and polars dataframes can't be mixed together, so the user has 
+    to switch between format if needed.
+    
+    Parameters
+    ----------
+    tables : list of tuples
+        List of (dataframe, columns_to_join, columns_to_agg) tuple
+        specifying the auxilliary dataframes and their columns for joining 
+        and aggregation operations.
+    
+        dataframe : {pandas.DataFrame, polars.DataFrame, polars.LazyFrame}
+            The auxilliary data to aggregate and join.
+        
+        columns_to_join : str or array-like
+            Select the columns from the dataframe to use as keys during the join operation.
+        
+        columns_to_agg : str or array-like
+            Select the columns from the dataframe to use as values during 
+            the aggregation operations.
+
+    main_key : str or array-like
+        Select the columns from the base table to use as keys during 
+        the join operation.
+
+    agg_ops : str or list of str, default=None
+        Aggregation operations to perform on the auxilliary table.
+        Options: {'mean', 'std', 'min', 'max', 'mode'}. If set to None, 
+        ['mean', 'mode'] will be used.
+    """
+
+    def __init__(self, tables, main_key, agg_ops=None):
+        self.tables = tables
+        self.main_key = main_key
+        self.agg_ops = agg_ops
+        self.assembly_engine = AssemblingEngine.get_for(tables)
```

Comment:
I would be prefer if this was done in the fit, as it would be legit to change the tables attribute.

In general the scikit-learn guidelines are to do as little as possible in the init, which only serves to store the configuration of the object, and move as much as possible to the fit.

Reply:
I had in mind [this design pattern from ArgKmin in scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp#L343), but point taken let's change this